### ISSUE TITLE
Validate implementation with official NIST Tech Pubs collection

### ISSUE THREAD

**Issue by ronaldtse** [metanorma/nist-pubid#8]
https://github.com/usnistgov/NIST-Tech-Pubs/blob/nist-pages/xml/allrecords.xml

Data e.g.

```xml
<body>
   <query key="BH">
      <doi type="report-paper_title">10.6028/NBS.BH.1</doi>
      <crm-item name="publisher-name" type="string">National Institute of Standards and Technology (NIST)</crm-item>
      <crm-item name="prefix-name" type="string">National Institute of Standards and Technology</crm-item>
      <crm-item name="member-id" type="number">4068</crm-item>
      <crm-item name="citation-id" type="number">78696207</crm-item>
      <crm-item name="book-id" type="number">2050209</crm-item>
      <crm-item name="deposit-timestamp" type="number">201511031134</crm-item>
      <crm-item name="owner-prefix" type="string">10.6028</crm-item>
      <crm-item name="last-update" type="date">2018-03-06T09:55:24Z</crm-item>
      <crm-item name="created" type="date">2015-11-04T17:31:05Z</crm-item>
      <crm-item name="citedby-count" type="number">0</crm-item>
      <doi_record>
         <report-paper>
            <report-paper_metadata language="en">
               <contributors>
                  <person_name sequence="first" contributor_role="author">
                     <given_name>Ira H</given_name>
                     <surname>Woolson</surname>
                  </person_name>
<!-- ... -->
               </contributors>
               <titles>
                  <title>Recommended minimum requirements for small dwelling construction :</title>
                  <subtitle>report of Building Code Committee July 20, 1922</subtitle>
               </titles>
               <edition_number>0</edition_number>
               <publication_date media_type="online">
                  <year>1923</year>
               </publication_date>
               <publisher>
                  <publisher_name>National Bureau of Standards</publisher_name>
                  <publisher_place>Gaithersburg, MD</publisher_place>
               </publisher>
               <institution>
                  <institution_name>National Bureau of Standards</institution_name>
                  <institution_acronym>NBS</institution_acronym>
                  <institution_place>Gaithersburg, MD</institution_place>
               </institution>
               <publisher_item>
                  <item_number item_number_type="report-number">NBS BH 1</item_number>
               </publisher_item>
               <doi_data>
                  <doi>10.6028/NBS.BH.1</doi>
                  <resource>https://nvlpubs.nist.gov/nistpubs/Legacy/BH/nbsbuildinghousing1.pdf</resource>
               </doi_data>
            </report-paper_metadata>
         </report-paper>
      </doi_record>
   </query>
```

We want to be able to generate a PubID for every entry, using a combination of:
* `<item_number>`
*  `<edition_number>`
* `<publication_date>` `<year/month>`
* `<doi>`

Right now that information is handled in https://github.com/relaton/relaton-nist/blob/master/lib/relaton_nist/nist_bibliography.rb

And create a map.

**Comment by ronaldtse** [comment]
For @mico

**Comment by ronaldtse** [comment]
From @mico 

> I need more data for Section, Supplement, Index, Insert, Errata. Examples and more details.

**Comment by mico** [comment]
I can validate implementation by parsing `<item_number>` and trying to convert it to short code and comparing with original `<item_number>`. Is that how you imagine validation process?

Do I understand correctly that `<item_number>` is already NIST PubID?

As I can see, usually it includes everything, but sometimes `<edition_number>` is not included, examples:
```
 <doi_data>
    <doi>10.6028/NIST.IR.8216</doi>
    <resource>https://nvlpubs.nist.gov/nistpubs/ir/2018/NIST.IR.8216.pdf</resource>
 </doi_data>
 <publisher_item>
    <item_number item_number_type="report-number">NIST IR 8216</item_number>
 </publisher_item>
 <edition_number>1.1</edition_number>
```

another doc, edition included:
```
 <edition_number>Revision 5</edition_number>
 <publisher_item>
    <identifier id_type="report-number">NIST SP 800-53r5</identifier>
 </publisher_item>
 <doi_data>
    <doi>10.6028/NIST.SP.800-53r5</doi>
    <resource>https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-53r5.pdf</resource>
 </doi_data>

```

> We want to be able to generate a PubID for every entry, using a combination of:
> 
> * `<item_number>`
> * `<edition_number>`
> * `<publication_date>` `<year/month>`
> * `<doi>`

You mean we want to be able to parse and convert data using this fields, right?
We can convert using `<item_number>` or `<doi>` + `<edition_number>` and `<publication_date>` `<year/month>` together, because they are not included to `<item_number>` or `<doi>`.

**Comment by ronaldtse** [comment]
> I can validate implementation by parsing `<item_number>` and trying to convert it to short code and comparing with original `<item_number>`. Is that how you imagine validation process?

This process sounds right. Instead of a "validation", let's do a check on how many PubIDs are different from the originals. This will be useful for NIST in the migration to the PubID.

> Do I understand correctly that `<item_number>` is already NIST PubID?

Unfortunately, it is not. The current `<item_number>` is a manually assigned identifier with some consistency but not fully consistent like the PubID. That's why they wish to move to a defined scheme.

> You mean we want to be able to parse and convert data using this fields, right? We can convert using `<item_number>` or `<doi>` + `<edition_number>` and `<publication_date>` `<year/month>` together, because they are not included to `<item_number>` or `<doi>`.

Yes. We want to generate a new PubID for every entry, and compare the "short" and "machine-readable" entries with the current `<item_number>` and `<doi>`.

This is the situation:
1. Every NIST Tech Pub has an "document identifier" and a "DOI identifier". They are not created in a consistent way.
2. In the new NIST PubID, there is a way to generate a consistent "document identifier" and a "machine-readable identifier" (which can be used like DOI).
3. NIST wants to migrate all their "document identifiers" and "DOI identifiers" to the new NIST PubID scheme.
4. We need to help them demonstrate the impact of this migration -- how many identifiers need to be changed, how large is the change, how many DOI identifiers need to change, what to change to, etc.

@mico can you help move towards this direction?

**Comment by mico** [comment]
@ronaldtse there is still a lot of work to parse all NIST Tech Pubs collection's pubids correctly.
Now my pubids list for parsing updates looks like:

(parsed result | original pubid from NIST Tech Pubs)
```
NBS CIRC 101supe2 | NBS CIRC 101e2sup
NBS CIRC 11e2 | NBS CIRC 11e2-1915
NBS CIRC 11e3 | NBS CIRC 11e3-1917
NBS CIRC 13e2 | NBS CIRC 13e2revjune1908
NBS CIRC 13sup1928e11 | NBS CIRC 13e11sup1928
NBS CIRC 13e2 | NBS CIRC 13e2revJune1908
NBS CIRC 154sup | NBS CIRC 154suprev
NBS CIRC 15 | NBS CIRC 15-April1909
NBS CIRC 24supe4 | NBS CIRC 24e4sup
NBS CIRC 24sup | NBS CIRC 24supJan1924
NBS CIRC 24sup | NBS CIRC 24supJuly1922
NBS CIRC 25sup | NBS CIRC 25sup-1924
NBS CIRC 25sup | NBS CIRC 25sup-1925
NBS CIRC 25sup | NBS CIRC 25sup-1927
NBS CIRC 25sup | NBS CIRC 25sup-1928
NBS CIRC 25sup | NBS CIRC 25sup-1929
NBS CIRC 488SEC | NBS CIRC 488sec1
NBS CIRC 488SEC | NBS CIRC 488sec2
NBS CIRC 488SEC | NBS CIRC 488sec3
NBS CIRC 488SEC | NBS CIRC 488sec4
NBS CIRC 488SEC | NBS CIRC 488sec5
NBS CRPL 1-2 | NBS CRPL 1-2_3-1
NBS CRPL 1-2 | NBS CRPL 1-2_3-1A
NBS CRPL 4 | NBS CRPL 4-m-5
NBS CRPL 8-1 | NBS CRPL 8-1_9-4
NBS HB 44e2 | NBS HB 44e2-1955
NBS HB 44e2 | NBS HB 44e2-1961
NBS IR 80-2073 | NBS IR 80-2073.2
NBS IR 80-2073 | NBS IR 80-2073.3
NBS MP 39 | NBS MP 39(1)
NBS MP 39 | NBS MP 39(2)
NBS MP 39 | NBS MP 39(3)
NBS RPT 8 | NBS RPT 8/19/1927
NBS SP 250e1982 | NBS SP 250e1982app
NIST HB 133e4 | NIST HB 133e4-2002
NIST HB 133e4 | NIST HB 133e4-2003
NIST HB 133e4 | NIST HB 133e4-2005
NIST SP 1011 | NIST SP 1011-I-2.0
NIST SP 1011 | NIST SP 1011-II-1.0
NIST SP 1075 | NIST SP 1075-NCNR
NIST SP 1113 | NIST SP 1113-BFRL
NIST SP 1113 | NIST SP 1113-PERMIS
NIST SP 250e1992 | NIST SP 250e1992app
NIST SP 260-126 | NIST SP 260-126 rev 2013
NIST SP 260-162e2006 | NIST SP 260-162 2006ed.
NIST SP 800-22-1a | NIST SP 800-22r1a
NIST SP 800-27RA-a | NIST SP 800-27ra
NIST SP 800-38A-a | NIST SP 800-38a-add
NIST SP 800-53AR | NIST SP 800-53ar1
NIST SP 800-57P-1 | NIST SP 800-57Pt3r1
NIST SP 801 | NIST SP 801-errata
NIST SP 955 | NIST SP 955 Suppl.
NIST AMS 300-8-1 | NIST AMS 300-8r1 (February 2021 update)
```

After I finish this list, I'm going to compare doi (it could bring much more work to do), sometimes doi have more complete information, for example:
```
<item_number item_number_type="report-number">NIST IR 8178</item_number>
<doi>10.6028/NIST.IR.8178port</doi>
```
(item_number don't have translation part here)

**Comment by mico** [comment]
Seems this documents is very different, but why they have same number and edition? Only different year, trying to figure out how to represent them in NIST PubID.

`NBS CIRC 11e2` https://nvlpubs.nist.gov/nistpubs/Legacy/circ/nbscircular11e2.pdf
`NBS CIRC 11e2-1915` https://nvlpubs.nist.gov/nistpubs/Legacy/circ/nbscircular11e2-1915.pdf

**Comment by ronaldtse** [comment]
Are these the Short PubID representations?

> NBS CIRC 101supe2 | NBS CIRC 101e2sup
> NBS CIRC 13sup1928e11 | NBS CIRC 13e11sup1928
> NBS CIRC 24supe4 | NBS CIRC 24e4sup

The first one refers to the Supplement of "NBS CIRC 101e2", so "NBS CIRC 101supe2" won't make sense. The others are similar.

> NBS CIRC 154sup | NBS CIRC 154suprev

This is a revision of the supplement. 

> NBS CIRC 13e2 | NBS CIRC 13e2revjune1908
> NBS CIRC 13e2 | NBS CIRC 13e2revJune1908

> NBS CIRC 15 | NBS CIRC 15-April1909

> NBS CIRC 24sup | NBS CIRC 24supJan1924
> NBS CIRC 24sup | NBS CIRC 24supJuly1922

These are two separate documents. The first is a supplement issued Jan 1924, the second is another supplement issued July 1922. The PubID must differentiate them.

> NBS CIRC 25sup | NBS CIRC 25sup-1924
> NBS CIRC 25sup | NBS CIRC 25sup-1925
> NBS CIRC 25sup | NBS CIRC 25sup-1927
> NBS CIRC 25sup | NBS CIRC 25sup-1928
> NBS CIRC 25sup | NBS CIRC 25sup-1929

Multiple supplements published in different years, need to distinguish them.

> NBS CIRC 488SEC | NBS CIRC 488sec1
> NBS CIRC 488SEC | NBS CIRC 488sec2
> NBS CIRC 488SEC | NBS CIRC 488sec3
> NBS CIRC 488SEC | NBS CIRC 488sec4
> NBS CIRC 488SEC | NBS CIRC 488sec5

Different "sections" of a document. e.g. https://nvlpubs.nist.gov/nistpubs/Legacy/circ/nbscircular488sec5.pdf

Need to reflect them in the PubID.

> NBS CRPL 1-2 | NBS CRPL 1-2_3-1
> NBS CRPL 8-1 | NBS CRPL 8-1_9-4

We need to find a way to identify these. Perhaps "1-2, 3-1" is the number?

<img width="991" alt="Screenshot 2021-12-23 at 11 55 06 AM" src="https://user-images.githubusercontent.com/11865/147185678-f362c574-07fc-46f5-a66c-1caaa6217e32.png">

> NBS CRPL 1-2 | NBS CRPL 1-2_3-1A

This is the "Supplement to report CRPL-1-2, 3-1"

> NBS CRPL 4 | NBS CRPL 4-m-5

This document is really called "CRPL 4-M-5":
<img width="1014" alt="Screenshot 2021-12-23 at 11 46 06 AM" src="https://user-images.githubusercontent.com/11865/147185080-69750fb3-74cc-437b-8877-4045c14c14d9.png">

> NBS HB 44e2 | NBS HB 44e2-1955
> NBS HB 44e2 | NBS HB 44e2-1961

Two separate documents, need to distinguish them. 
<img width="564" alt="Screenshot 2021-12-23 at 11 57 13 AM" src="https://user-images.githubusercontent.com/11865/147185819-b9be9838-13a6-413b-a9be-1182481d888c.png">


> NBS IR 80-2073 | NBS IR 80-2073.2
> NBS IR 80-2073 | NBS IR 80-2073.3

These are different parts.
<img width="828" alt="Screenshot 2021-12-23 at 11 57 39 AM" src="https://user-images.githubusercontent.com/11865/147185846-1c83fb3e-7c69-46c4-9239-ad52bad2ad2c.png">
<img width="867" alt="Screenshot 2021-12-23 at 11 57 57 AM" src="https://user-images.githubusercontent.com/11865/147185870-147d9d50-5087-4b12-81ff-55165ccbe881.png">


> NBS MP 39 | NBS MP 39(1)
> NBS MP 39 | NBS MP 39(2)
> NBS MP 39 | NBS MP 39(3)

These are different editions of NBS MP 39.

<img width="761" alt="Screenshot 2021-12-23 at 11 59 06 AM" src="https://user-images.githubusercontent.com/11865/147185958-5b1e45f6-314d-45e4-ade8-f1f57c27cde8.png">
<img width="738" alt="Screenshot 2021-12-23 at 11 59 26 AM" src="https://user-images.githubusercontent.com/11865/147185981-9d6bb908-bf5b-4948-ae94-bb87e0ddca62.png">


> NBS RPT 8 | NBS RPT 8/19/1927

The original ID is wrong. This is actually:
```
Report Number: NBS RPT 8/19/1927
doi:10.6028/NBS.RPT.10093
```
=> NBS RPT 10093

<img width="900" alt="Screenshot 2021-12-23 at 11 48 13 AM" src="https://user-images.githubusercontent.com/11865/147185262-79ee249c-eff3-4aaf-8085-bf72663276d1.png">

> NBS SP 250e1982 | NBS SP 250e1982app
> NIST SP 250e1992 | NIST SP 250e1992app

This is the "appendix" of NBS SP 250e1982. So "Appendix" needs to be an additional element.

> NIST HB 133e4 | NIST HB 133e4-2002
> NIST HB 133e4 | NIST HB 133e4-2003
> NIST HB 133e4 | NIST HB 133e4-2005

Same edition, but different publication year, means that the "edition" number and the "publication year" are both significant in identification. The publication year means that there were "amendments incorporated as of the publication year".

2002 edition:

<img width="1041" alt="Screenshot 2021-12-23 at 11 49 43 AM" src="https://user-images.githubusercontent.com/11865/147185370-279f6f4d-a8f7-4540-a54d-d79d70221007.png">

<img width="849" alt="Screenshot 2021-12-23 at 11 50 04 AM" src="https://user-images.githubusercontent.com/11865/147185396-a092cb8c-96f6-4e2a-8efb-5f6bf261c253.png">

2005 edition:
<img width="1054" alt="Screenshot 2021-12-23 at 11 51 05 AM" src="https://user-images.githubusercontent.com/11865/147185460-3d8edd14-4f8b-45b9-9a65-e49edce82973.png">
<img width="581" alt="Screenshot 2021-12-23 at 11 51 37 AM" src="https://user-images.githubusercontent.com/11865/147185484-ea1b7f5d-b141-41cd-bee0-c8f9a401336c.png">


> NIST SP 1011 | NIST SP 1011-I-2.0
> NIST SP 1011 | NIST SP 1011-II-1.0

Two parts: "I" and "II", each with a different version number.

> NIST SP 1075 | NIST SP 1075-NCNR

This is described in relaton/relaton-nist#53 and usnistgov/NIST-Tech-Pubs#5.

In the NIST Tech Pubs history it rarely happens, but this SP document number is duplicated. 

> NIST SP 1113 | NIST SP 1113-BFRL
> NIST SP 1113 | NIST SP 1113-PERMIS

This is another case of a duplicated ID.

`NIST SP 1113-BFRL`: (July 2010)
NIST Special Publication 1113. Real-Time Monitoring of Total Inward Leakage of Respiratory Equipment
<img width="660" alt="Screenshot 2021-12-23 at 12 08 51 PM" src="https://user-images.githubusercontent.com/11865/147186672-bf7f2680-f627-4f57-a5e2-15306f12572c.png">



`NIST SP 1113-PERMIS`: (September 2010)
<img width="1011" alt="Screenshot 2021-12-23 at 12 08 09 PM" src="https://user-images.githubusercontent.com/11865/147186633-ffd345c6-d1d5-4387-a602-f1ab24575ffe.png">


> NIST SP 260-126 | NIST SP 260-126 rev 2013

This is "Revision 2013 of NIST SP 260-126".

<img width="918" alt="Screenshot 2021-12-23 at 12 24 41 PM" src="https://user-images.githubusercontent.com/11865/147187676-5f7b1d0f-f497-4ac7-9baa-46a5b1c4b94c.png">

<img width="941" alt="Screenshot 2021-12-23 at 12 24 34 PM" src="https://user-images.githubusercontent.com/11865/147187662-0e52e73a-a9be-43b7-8114-7652c76bd650.png">

> NIST SP 260-162e2006 | NIST SP 260-162 2006ed.

This is just "NIST Special Publication 260-162" Edition 1.

<img width="893" alt="image" src="https://user-images.githubusercontent.com/11865/147187766-d6a9c213-7471-45fd-b810-fddd584a7649.png">

> NIST SP 800-22-1a | NIST SP 800-22r1a

This is "SP 800-22 Rev. 1a" or "NIST SP 800-22rev1a".

> NIST SP 800-27RA-a | NIST SP 800-27ra

This is "NIST Special Publication 800-27 Rev A."

> NIST SP 800-38A-a | NIST SP 800-38a-add

This is the Addendum to "NIST SP 800-38A" ("A" here means "Part A").

> NIST SP 800-53AR | NIST SP 800-53ar1

This is "NIST Special Publication 800-53A Revision 1".

> NIST SP 800-57P-1 | NIST SP 800-57Pt3r1

This is "NIST SP 800-57 Part 3 Revision 1".

> NIST SP 801 | NIST SP 801-errata

"Errata for NIST SP 801".

> NIST SP 955 | NIST SP 955 Suppl.

"Supplement to NIST SP 955".

> NIST AMS 300-8-1 | NIST AMS 300-8r1 (February 2021 update)

This is a Revision 1 of AMS 300-8. The parsed PubID now puts the "1" as a "part" which is wrong.

**Comment by mico** [comment]
> This is a Revision 1 of AMS 300-8. The parsed PubID now puts the "1" as a "part" which is wrong.
After looking at these examples in https://www.nist.gov/system/files/documents/2021/08/26/Publication-ID-Proposal_26Aug21.pdf I made it work this way.
<img width="672" alt="image" src="https://user-images.githubusercontent.com/5849/147208605-fa28a584-4534-4f0d-8332-7c7cc65a95c0.png">
I got to conclusion: If there are no other parts (volume, part, edition) then PubID with revision should look like: `NIST AMS 300-8-1`.
But if there are other part, then revision with letter will be added, example: `NIST AMS 300-8v1r1`

**Comment by ronaldtse** [comment]
> I got to conclusion: If there are no other parts (volume, part, edition) then PubID with revision should look like: `NIST AMS 300-8-1`. But if there are other part, then revision with letter will be added, example: `NIST AMS 300-8v1r1`

Interesting. Did not know sometimes the edition/revision is represented as "-N". I think it will be very confusing to use "-N" as edition vs "rN" (e.g revision 4 as "r4"). We will have to report back to NIST that this is an issue.

**Comment by mico** [comment]
> > I got to conclusion: If there are no other parts (volume, part, edition) then PubID with revision should look like: `NIST AMS 300-8-1`. But if there are other part, then revision with letter will be added, example: `NIST AMS 300-8v1r1`
> 
> Interesting. Did not know sometimes the edition/revision is represented as "-N". I think it will be very confusing to use "-N" as edition vs "rN" (e.g revision 4 as "r4"). We will have to report back to NIST that this is an issue.

Should I change it back to `NIST AMS 300-8r1` instead of `NIST AMS 300-8-1` before we get response from NIST?

**Comment by ronaldtse** [comment]
Yes, I would prefer using "r1" instead of "-1" as it is unambiguous.

**Comment by mico** [comment]
Another issue, in the https://www.nist.gov/system/files/documents/2021/08/26/Publication-ID-Proposal_26Aug21.pdf proposed to use form:
`{series}({stage}) {report number}{part}{edition}({translation})` where
`{part}` including supplement, if we follow this logic we are getting `NBS CIRC 101supe2` instead of `NBS CIRC 101e2sup`.
maybe I should move supplement as separate entity, like:
`{series}({stage}) {report number}{part}{edition}{supplement}({translation})`.
@ronaldtse do you agree?

**Comment by ronaldtse** [comment]
Yes, Supplement, Addendum and Index etc should be considered "localities" instead of parts. 

Isn't there another case where there are supplements issued in different years too? Then we need to consider the numbering of supplements as well.

**Comment by mico** [comment]
> Isn't there another case where there are supplements issued in different years too? Then we need to consider the numbering of supplements as well.

Is it different supplements issues in different years:
https://nvlpubs.nist.gov/nistpubs/Legacy/circ/nbscircular25sup-1924.pdf and https://nvlpubs.nist.gov/nistpubs/Legacy/circ/nbscircular25sup-1925.pdf ?

**Comment by ronaldtse** [comment]
Yes those were the documents I meant. My problem is that the identifier “NBS CIRC 25-1924sup” does not seem to distinguish between:
- NBS Circular 25 Supplement 1924
- NBS Circular 25 Edition 1924 Supplement 1

Or is this already clear?

Can a Supplement have its own edition? Maybe not?

**Comment by mico** [comment]
> * NBS Circular 25 Supplement 1924

"NBS CIRC 25-1924sup"

> * NBS Circular 25 Edition 1924 Supplement 1

Can be "NBS CIRC 25e1924sup"?

It's not clear in proposal where "edition" should be. There is option to represent it as "NBS CIRC 25e1924sup" as well as "NBS CIRC 25-1924sup".
Maybe "edition" in the sequence number (report number) is not the same as "edition" after "e" character?

**Comment by ronaldtse** [comment]
> > * NBS Circular 25 Supplement 1924
> 
> "NBS CIRC 25-1924sup"

This should be `NBS CIRC 25sup1924`.

> > * NBS Circular 25 Edition 1924 Supplement 1
> 
> Can be "NBS CIRC 25e1924sup"?

This should be `NBS CIRC 25-1924sup1`.

See image for reference:

<img width="545" alt="Screenshot 2022-01-03 at 2 02 18 PM" src="https://user-images.githubusercontent.com/11865/147903111-47c23aee-732e-4386-b70d-e2939ca0fcb3.png">

**Comment by mico** [comment]
> * NBS Circular 25 Supplement 1924
> * NBS Circular 25 Edition 1924 Supplement 1

I didn't find the second document, is it exist? If not, could you find a real example for this problem?

**Comment by ronaldtse** [comment]
@mico it was a fictitious example, can you try finding one example in `allrecords.xml`? Even if it does not exist today, it may exist in the future in the current rules, so we have to handle it.

**Comment by mico** [comment]
> This is the "appendix" of NBS SP 250e1982. So "Appendix" needs to be an additional element.

We don't have appendix in PubID proposal, should they update it?

**Comment by ronaldtse** [comment]
> We don't have appendix in PubID proposal, should they update it?
> Should it be "NBS SP 250e1982app" when we create PubID or "NBS SP 250e1982A" or something else?

@mico yes, can you file this as a separate issue so we can report it? Thanks.

We can't use "A" because it is commonly used as a Part identifier. Maybe "app" is appropriate.
